### PR TITLE
JDK13 bringup - adding pConfig JAVA13

### PR DESCRIPTION
**Pull request by JasonFengJ9**
`JDK13` bringup - adding `pConfig JAVA13`

1. added `pConfig JAVA13` along with flag `Java13`;
2. added a field `java.lang.invoke.InvokerBytecodeGenerator.HIDDEN_SIG` for `JDK13`;
3. added a build flag `ive_rawBuild` which is not enabled;
4. removed `J2SE_SHAPE` related code;
5. use `ifdef J9VM_IVE_RAW_BUILD` in place of `J2SE_SHAPE(vm) == J2SE_SHAPE_RAW`.

Manually verified that `JDK13` can be built and `-version` output is:
```
openjdk version "13-internal" 2019-09-17
OpenJDK Runtime Environment (build 13-internal+0-adhoc.jason.openj9-openjdk-jdk)
Eclipse OpenJ9 VM (build master-9d606fc, JRE 13 Linux amd64-64-Bit Compressed References 20190225_000000 (JIT enabled, AOT enabled)
OpenJ9   - 9d606fc
OMR      - e6c6285
JCL      - b0f7ca0 based on jdk-13+6)
```

related: https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/81, https://github.com/ibmruntimes/openj9-openjdk-jdk8/pull/273
closes: https://github.com/eclipse/openj9/issues/4857

Reviewer: @DanHeidinga 

Signed-off-by: Jason Feng <fengj@ca.ibm.com>

### CODE REVIEW

File: runtime/j9vm/jvm.c [hunk 1]
```diff
@@ -1469,18 +1471,14 @@ getVersionFromReleaseFile(void)
 	if (NULL != propsFile) {
 		const char *version = props_file_get(propsFile, "JAVA_VERSION");
```
Comment (pshipton) [inline runtime/j9vm/jvm.c]: Not sure any of this is useful, we build specifically for the openjdk version. How about deleting this code to make future ports that must easier.

File: runtime/oti/j2sever.h [hunk 2]
```diff
@@ -43,8 +44,12 @@
 	#define J2SE_LATEST  J2SE_18
 #elif JAVA_SPEC_VERSION == 11
 	#define J2SE_LATEST  J2SE_V11
-#else
+#elif JAVA_SPEC_VERSION == 12
 	#define J2SE_LATEST  J2SE_V12
+#elif JAVA_SPEC_VERSION == 13
+	#define J2SE_LATEST  J2SE_V13
```
Comment (pshipton) [inline runtime/oti/j2sever.h]: How about `#define J2SE_LATEST (JAVA_SPEC_VERSION << 16)`

File: runtime/oti/j2sever.h [hunk 3]
```diff
@@ -70,12 +75,15 @@
 	#define J2SE_SHAPE_LATEST       J2SE_SHAPE_V11
 #elif JAVA_SPEC_VERSION == 12
 	#define J2SE_SHAPE_LATEST       J2SE_SHAPE_V12
+#elif JAVA_SPEC_VERSION == 13
+	#define J2SE_SHAPE_LATEST       J2SE_SHAPE_V13
```
Comment (pshipton) [inline runtime/oti/j2sever.h]: J2SE_SHAPE_ doesn't seem to be used other than J2SE_SHAPE_RAW. Can we get rid of these constants?

Comment (JasonFengJ9) [reply runtime/oti/j2sever.h]: Sure.

File: runtime/oti/j2sever.h (hunk 2 above)
Comment (JasonFengJ9) [reply runtime/oti/j2sever.h]: Agreed.

File: runtime/j9vm/jvm.c (hunk 1 above)
Comment (JasonFengJ9) [reply runtime/j9vm/jvm.c]: Yeah, it can be simpler though we need consider the impact on legacy non-openj9 Java 8 builds.

Comment (pshipton) [reply runtime/j9vm/jvm.c]: I thought that uses classlib.properties?

Comment (pshipton) [reply runtime/j9vm/jvm.c]: Seems getVersionFromReleaseFile() and the caller can be deleted and it would be fine.

**Comment by JasonFengJ9** [comment]
@pshipton the PR has been updated accordingly, please review.

File: runtime/j9vm/jvm.c [hunk 4]
```diff
@@ -5515,7 +5392,8 @@ JVM_InitClassName
 
 	Trc_SC_GetClassName_Entry(env, theClass);
 
-	if (J2SE_SHAPE_RAW == J2SE_SHAPE(vm)) {
+#ifdef J9VM_IVE_RAW_BUILD /* J9VM_IVE_RAW_BUILD is not enabled */
```
Comment (pshipton) [inline runtime/j9vm/jvm.c]: J9VM_IVE_RAW_BUILD is not enabled *by default* 
is better

File: runtime/j9vm/jvm.c [hunk 5]
```diff
@@ -1370,157 +1355,49 @@ decodeSetting(const char* key, const char* value, const VersionSetting* settings
...
-#ifdef DEBUG
-			printf("No 'version' property in %s\n", jvmBufferData(propsPathBuffer));
-#endif
-			goto bail;
-		}
+	if (-1 == jvmSEVersion) {
+		PORT_ACCESS_FROM_PORT(&j9portLibrary);
+		J9StringBuffer *propsPathBuffer = NULL;
+		j9props_file_t propsFile = NULL;
+		UDATA finalVersion = J2SE_18;
```
Comment (pshipton) [inline runtime/j9vm/jvm.c]: Why this default? I don't see this behavior in the original code.

Comment (JasonFengJ9) [reply runtime/j9vm/jvm.c]: When there is a `classlib.properties`, default version is `J2SE_18` for `Java 8` which is the only build with this file.
Existing code sets the `finalVersion` to `J2SE_LATEST` if there is an error.
I thought this makes it is explicit that only non-openj9 `Java 8` build has this property file.

File: runtime/j9vm/jvm.c (hunk 4 above)
Comment (JasonFengJ9) [reply runtime/j9vm/jvm.c]: sure.

File: runtime/j9vm/jvm.c (hunk 5 above)
Comment (pshipton) [reply runtime/j9vm/jvm.c]: I think it makes more sense to set J2SE_LATEST as the default. Just because a classlib.properties file is found doesn't mean it should default to J2SE_18 when the file doesn't contain a recognized version.  Defaulting to J2SE_LATEST is how it works now. J2SE_LATEST will be J2SE_18 for any Java 8 build anyway.

Also, I'm fine to have J2SE_LATEST as the initialization value for finalVersion, it doesn't need to be set in all the cases.

Comment (JasonFengJ9) [reply runtime/j9vm/jvm.c]: > J2SE_LATEST will be J2SE_18 for any Java 8 build anyway.

It seems there is no need to read/parse `classlib.properties`, i.e., always use `J2SE_LATEST`. The property file can be deleted.

Comment (pshipton) [reply runtime/j9vm/jvm.c]: We can't delete the property file, it is used for other things like the bootclasspath. However good point, we don't need to read the version.

**Comment by pshipton** [comment]
Jason, as part of this change, can you please change the Panama JPP config to depend on JAVA13. As discussed with @tajila Panama needs to be on the latest.

File: runtime/j9vm/jvm.c (hunk 5 above)
Comment (JasonFengJ9) [reply runtime/j9vm/jvm.c]: PR has been updated, please review.

File: runtime/classlib.properties [hunk 6]
```diff
@@ -1,7 +1,29 @@
...
+################################################################################
+
+# Properties file for Java 8 (Java 11 and beyond doesn't need this file)
 
 # Shape of the class library (e.g. Sun.Basic, Harmony)
 shape=vm.shape
 
 # Version of the class library
-version=1.9
+version=1.8
```
Comment (pshipton) [inline runtime/classlib.properties]: What is this file used for? It doesn't define the bootclasspath.

File: runtime/oti/j2sever.h [hunk 7]
```diff
@@ -27,60 +27,24 @@
...
-#define J2SE_SHAPE_OPENJDK 		0x10000
-#define J2SE_SHAPE_V11			0x80000
-#define J2SE_SHAPE_V12			0x90000
-#define J2SE_SHAPE_RAWPLUSJ9	0xA0000
-#define J2SE_SHAPE_RAW	 		0xB0000
-#define J2SE_SHAPE_MASK 		0xF0000
-#define J2SE_SHAPE_SHIFT		16
 #define J2SE_JAVA_SPEC_VERSION_SHIFT 8
+/* J2SE_LATEST is the highest Java version supported by VM for a JCL level. */
+#define J2SE_LATEST (JAVA_SPEC_VERSION << J2SE_JAVA_SPEC_VERSION_SHIFT)
```
Comment (pshipton) [inline runtime/oti/j2sever.h]: This isn't well named, how about J2SE_CURRENT_VERSION

File: runtime/j9vm/jvm.c [hunk 8]
```diff
@@ -1802,7 +1629,7 @@ jint JNICALL JNI_CreateJavaVM(JavaVM **pvm, void **penv, void *vm_args) {
 	UDATA argEncoding = ARG_ENCODING_DEFAULT;
 	UDATA altJavaHomeSpecified = 0; /* not used on non-Windows */
 	J9PortLibraryVersion portLibraryVersion;
-	UDATA j2seVersion;
+	UDATA j2seVersion = J2SE_LATEST;
```
Comment (pshipton) [inline runtime/j9vm/jvm.c]: j2seVersion isn't really needed since you removed most of the references, you just need to fix the last one.

File: runtime/classlib.properties (hunk 6 above)
Comment (JasonFengJ9) [reply runtime/classlib.properties]: Please see https://github.com/ibmruntimes/openj9-openjdk-jdk8/pull/273.
`Java 8` extension hardcoded `-e 's/shape=vm.shape/shape=sun_openjdk\nbootpath=rt.jar:resources.jar:jsse.jar:charsets.jar:jce.jar/g'`.

Comment (pshipton) [reply runtime/classlib.properties]: Just seems weird that the only purpose of the file in OpenJ9 is so the extensions can modify it.

**Comment by pshipton** [comment]
Please rebase and fix the merge conflict.

File: runtime/classlib.properties (hunk 6 above)
Comment (JasonFengJ9) [reply runtime/classlib.properties]: Probably it was for customization accordingly to Java version, now `Java 8` is the only user. `
bootpath` can be hardcoded here within `OpenJ9` and the `Java 8` extension script can just copy it without modification.

Comment (pshipton) [reply runtime/classlib.properties]: Technically the extensions owns the bootclasspath, as the extensions define and create the jar files. If it were to be changed, I'd remove the file from OpenJ9 and create it entirely from the extensions.

File: runtime/j9vm/jvm.c [hunk 9]
```diff
@@ -1369,164 +1354,6 @@ decodeSetting(const char* key, const char* value, const VersionSetting* settings
...
-			goto bail;
-		}
-	
-		decoded = decodeSetting("shape", shape, SHAPE_SETTINGS, NUM_SHAPE_SETTINGS);
-		if (0 == decoded) {
-			goto bail;
-		}
-		finalVersion |= decoded;
-	
-		decoded = decodeSetting("version", version, VERSION_SETTINGS, NUM_VERSION_SETTINGS);
```
Comment (pshipton) [inline runtime/j9vm/jvm.c]: Is there any remaining use of decodeSetting()? Seem the impl can be deleted. Same with VERSION_SETTINGS.

File: runtime/jcl/common/jclcinit.c [hunk 10]
```diff
@@ -116,6 +116,13 @@ jint computeFullVersionString(J9JavaVM* vm)
 			j2se_version_info = "12.?";
 		}
 		break;
+	case J2SE_V13:
+		if ((J2SE_VERSION(vm) & J2SE_RELEASE_MASK) == J2SE_V13) {
+			j2se_version_info = "13";
```
Comment (pshipton) [inline runtime/jcl/common/jclcinit.c]: The next step would be to build this string automatically from the version, but I suggest we leave that to another PR.

File: runtime/classlib.properties (hunk 6 above)
Comment (JasonFengJ9) [reply runtime/classlib.properties]: In that case, I would like leave such change to another PR and allow this PR merge to unblock `JDK13` build.

File: runtime/j9vm/jvm.c (hunk 9 above)
Comment (JasonFengJ9) [reply runtime/j9vm/jvm.c]: Agreed, will remove it.

**Comment by pshipton** [comment]
jenkins test sanity plinux jdk8,jdk12,jdk

**Comment by pshipton** [comment]
jenkins tests sanity jdk

**Comment by pshipton** [comment]
jenkins tests sanity plinux jdk

**Comment by pshipton** [comment]
jenkins test sanity plinux jdk

**Comment by pshipton** [comment]
jenkins test sanity plinux jdk8

**Comment by pshipton** [comment]
jenkins test sanity xlinux jdk

**Comment by pshipton** [comment]
jenkins test sanity plinux jdknext

**Comment by pshipton** [comment]
jenkins test sanity xlinux jdknext

**Comment by pshipton** [comment]
jenkins test compile plinux jdknext

**Comment by pshipton** [comment]
jenkins test compile xlinux jdknext

**Comment by pshipton** [comment]
jenkins compile plinux jdknext

**Comment by pshipton** [comment]
jenkins compile xlinux jdknext

**Comment by pshipton** [comment]
jenkins compile win jdk8

**Comment by pshipton** [comment]
jenkins compile xlinux jdknext ibmruntimes/openj9-openjdk-jdk#81

**Comment by pshipton** [comment]
jenkins test sanity plinux jdk8 depends ibmruntimes/openj9-openjdk-jdk8#273

**Comment by pshipton** [comment]
jenkins test sanity plinux jdk8

**Comment by pshipton** [comment]
Windows is failing to compile.

**Comment by pshipton** [comment]
jenkins test sanity plinux jdknext depends ibmruntimes/openj9-openjdk-jdk#81

**Comment by pshipton** [comment]
The jdk13 build failed as well for different reasons. I messed up the command to run the build and didn't include ibmruntimes/openj9-openjdk-jdk#81 but not sure that matters to the failure.

**Comment by pshipton** [comment]
jenkins compile win jdk8

**Comment by pshipton** [comment]
jenkins test sanity zlinux jdk11

**Comment by pshipton** [comment]
Cannot test jdknext in PR testing atm due to https://github.ibm.com/runtimes/infrastructure/issues/1508

**Comment by pshipton** [comment]
jenkins compile win jdk8

**Comment by pshipton** [comment]
jenkins test sanity zlinux jdknext depends [ibmruntimes/openj9-openjdk-jdk#81](https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/81)

**Comment by pshipton** [comment]
jenkins test sanity zlinux jdknext depends ibmruntimes/openj9-openjdk-jdk#81

**Comment by pshipton** [comment]
jenkins test sanity plinux jdknext depends [ibmruntimes/openj9-openjdk-jdk#81](https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/81)

**Comment by pshipton** [comment]
jenkins test sanity plinux jdknext depends ibmruntimes/openj9-openjdk-jdk#81

**Comment by pshipton** [comment]
Opened https://github.com/eclipse/openj9/issues/4936 for the jdknext failure to build the tests.
